### PR TITLE
support unbind/bind SF through sysfs

### PR DESCRIPTION
**Pull request by l8huang**
Add GetSFAuxDev() function to get the auxiliary device name of a SF by 'sfnum'.

Add DeploySF() function to unbinds the SF from the default config driver and bind it to the actual SF driver.

### CODE REVIEW

File: sysfs_linux.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
+package mlxdevm
+
+// The following operations are currently not supported by the 'mlxdevm' devlink:
+// - Retrieving the auxiliary device name of an SF port, e.g., mlx5_core.sf.3.
+// - Unbinding the SF port from the default configuration driver and binding it to the actual SF driver.
+// These operations must be performed through sysfs. While we are looking for
+// adding devlink support for them in the long term, these functions within this
+// file can be used to accomplish the task until that support is implemented.
+
```
Comment (adrianchiris) [inline sysfs_linux.go:9]: This is unrelated to mlxdevm-go lib it does not deal with sysfs.

**Review by adrianchiris — COMMENTED** [review]
This is a NACK from my side.

mlxdevm-go lib aims to provide go API to initeract with mlxdevm generic netlink family.
any additions should go on top...

File: sysfs_linux.go (hunk 1 above)
Comment (l8huang) [reply sysfs_linux.go:9]: Unbind/bind SF is a mandatory step to deploy SF,  does mlxdevm generic netlink family support it now?

The reason of adding these functions is every user of this lib need to write similar code to deploy SF port. If they shouldn't be part of mlxdevm-go lib, do you have some suggestion where should they go? 

CC @venuiyer

Comment (venuiyer) [reply sysfs_linux.go:9]: @adrianchiris understand your concern. I think we need to determine what is expected from the caller 
when using some fns from this lib. E.g. when the trust function is called (the other PR), the SF should
not be attached, should we do that part of the trust call itself or expect the caller to make sure it is
detached before calling the trust?

It'll be nice to have the bind/unbind part of some lib, else every user will need to the same since 
as Lei says binding/unbinding is part of the seq to use the SF.
https://docs.nvidia.com/doca/sdk/nvidia+bluefield+dpu+scalable+function+user+guide/index.html (step 3)

Admittedly, it is not via mlxdevm; so should we instead have his part in sriovnet lib?

**Comment by adrianchiris** [comment]
>when the trust function is called (the other PR), the SF should
not be attached, should we do that part of the trust call itself or expect the caller to make sure it is
detached before calling the trust?

when creating SF there is a certain flow one need to take.

at high level
1. create devlink port (sf flavour)
2. set port attributes and activate SF
3. configure SF (devlink) device parameters (e.g enable netdev, rdma etc...)
4. reload device

this lib does not aim to provide API for high-level sf creation/deletion

> Admittedly, it is not via mlxdevm; so should we instead have his part in sriovnet lib?

sriovnet may be a good candidate given we can do everything with devlink i.e in a non vendor-specific manner. (i am not sure if currently there are gaps WRT upstream kernel) 
as an example in upstream there is no use of `sf_cfg` driver (moreover it doesnt exist upstream). also there were different parameter names WRT enabling/disabling eth netdev for SF between devlink upstream and mlxdevm.
a different path is to either create a new lib with the knobs to manage (create/configure/delete) SF or extend this one to do so.

within NVIDIA we have implemented a module to take care of that within our SF network DRA (Dynamic resource allocation) driver for k8s, but thats internal for now. We decided to implement that within the project until we have a stable API and there are additional use-case for open sourcing / contributing it.

for start, internally we can share it/ create a separate lib if it makes sense.

**Comment by venuiyer** [comment]
> > when the trust function is called (the other PR), the SF should
> > not be attached, should we do that part of the trust call itself or expect the caller to make sure it is
> > detached before calling the trust?
> 
> when creating SF there is a certain flow one need to take.
> 
> at high level
> 
> 1. create devlink port (sf flavour)
> 2. set port attributes and activate SF
> 3. configure SF (devlink) device parameters (e.g enable netdev, rdma etc...)
> 4. reload device
> 
> this lib does not aim to provide API for high-level sf creation/deletion
> 
> > Admittedly, it is not via mlxdevm; so should we instead have his part in sriovnet lib?
> 
> sriovnet may be a good candidate given we can do everything with devlink i.e in a non vendor-specific manner. (i am not sure if currently there are gaps WRT upstream kernel) as an example in upstream there is no use of `sf_cfg` driver (moreover it doesnt exist upstream). also there were different parameter names WRT enabling/disabling eth netdev for SF between devlink upstream and mlxdevm. a different path is to either create a new lib with the knobs to manage (create/configure/delete) SF or extend this one to do so.
> 
> within NVIDIA we have implemented a module to take care of that within our SF network DRA (Dynamic resource allocation) driver for k8s, but thats internal for now. We decided to implement that within the project until we have a stable API and there are additional use-case for open sourcing / contributing it.
> 
> for start, internally we can share it/ create a separate lib if it makes sense.

thanks @adrianchiris so if i understand we have internal lib to manage SF,  does that also include setting trust on the 
SF? Could you please share it.. i suppose we would have to include the lib as part of ovn-k8s till you are able
to have a stable API.

**Comment by adrianchiris** [comment]
> thanks @adrianchiris so if i understand we have internal lib to manage SF, does that also include setting trust on the
SF? Could you please share it.. i suppose we would have to include the lib as part of ovn-k8s till you are able
to have a stable API.

it does not. trust SF still need to be added to mlxdevm-go then extend the module which handles SF creation.
ill send the info via email.

**Comment by l8huang** [comment]
this change doesn't belong to this repo
